pkg/filtering/evaluate.go

```diff
@@ -0,0 +1,138 @@
+// Copyright (c) 2021 6 River Systems
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy of
+// this software and associated documentation files (the "Software"), to deal in
+// the Software without restriction, including without limitation the rights to
+// use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+// the Software, and to permit persons to whom the Software is furnished to do so,
+// subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+// FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+// COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+// IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+// CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+package filter
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+)
+
+// Evaluator interface which all grammar components should implement
+type Evaluator interface {
+	Evaluate(attrs map[string]string) (bool, error)
+}
+
+// Evaluate implementation for Basic Expression
+func (e *basicExpression) Evaluate(attrs map[string]string) (bool, error) {
+	switch {
+	case e.Has != nil:
+		return e.Has.Evaluate(attrs)
+	case e.Value != nil:
+		return e.Value.Evaluate(attrs)
+	case e.Predicate != nil:
+		return e.Predicate.Evaluate(attrs)
+	default:
+		return false, errors.New("unpopulated BasicExpression")
```

Comment:
Please follow the log message standard we have across the code base. (component: Error occured) and contextual fields.

Reply:
Declare the error beforehand just like we have done it elsewhere.

Reply:
Done